controllers/iamrole_controller.go

```diff
@@ -187,12 +204,35 @@ func (r *IamroleReconciler) HandleReconcile(ctx context.Context, req ctrl.Reques
 
 		}
 
-		if validation.CompareRole(ctx, *input, targetRole, *targetPolicy) {
-			log.Info("No change in the incoming policy compare to state of the world(external AWS IAM) policy")
-			r.UpdateStatus(ctx, iamRole, iammanagerv1alpha1.IamroleStatus{RetryCount: 0, RoleName: roleName, ErrorDescription: "", RoleID: aws.StringValue(targetRole.Role.RoleId), RoleARN: aws.StringValue(targetRole.Role.Arn), LastUpdatedTimestamp: iamRole.Status.LastUpdatedTimestamp, State: iammanagerv1alpha1.Ready})
+		if shouldManageSa {
+			// Ensure that the ServiceAccount is configured correctly. In the "Ready" handler here, we are getting the
+			// RoleARN from the Iamrole status itself.
+			saRequest := k8s.ServiceAccountRequest{
+				Namespace:          iamRole.Namespace,
+				IamRoleARN:         iamRole.Status.RoleARN,
+				ServiceAccountName: saName,
+			}
+			if _, err := rbac.EnsureServiceAccount(ctx, saRequest); err != nil {
+				log.Error(err, "error in updating service account for IRSA role")
+				r.Recorder.Event(iamRole, v1.EventTypeWarning, string(iammanagerv1alpha1.Error), "Unable to create/update service account for IRSA role due to error "+err.Error())
+				return r.UpdateStatus(ctx, iamRole, iammanagerv1alpha1.IamroleStatus{RetryCount: iamRole.Status.RetryCount + 1, RoleName: roleName, ErrorDescription: err.Error(), State: iammanagerv1alpha1.Error, LastUpdatedTimestamp: metav1.Now()}, requeueTime)
+			}
+		}
 
+		if !validation.CompareRole(ctx, *input, targetRole, *targetPolicy) {
+			shouldReconcile = true
+			r.Recorder.Event(iamRole, v1.EventTypeWarning, string(iammanagerv1alpha1.Error), "Detected difference between desired IAM policy and existing policy")
+		} else {
+			log.V(1).Info("No change in the incoming policy compare to state of the world(external AWS IAM) policy")
+		}
+
+		if !shouldReconcile {
```

Comment:
can this block just move inside the `if !validation.CompareRole` block above and then get rid of the `shouldReconcile` boolean? I much prefer to avoid boolean "temporary state" like that if I can, I find it very confusing

Reply:
Yeah - it was there originally, but I actually missed it and had logic _after_ it that was never being executed.. so I thought this was somehow cleaner. It probably isn't though. I'll revert that part.